File: app/src/main/java/com/filtertube/app/data/MixBuilder.kt
package com.filtertube.app.data

import kotlin.math.pow

/**
 * רשימת השמעה שנבנתה אוטומטית מהטעם של המשתמש.
 *
 * @param seedChannel הערוץ שסביבו נבנה המיקס, אם יש — לצורך תמונת הכריכה.
 */
data class Mix(
    val id: String,
    val title: String,
    val subtitle: String,
    val songs: List<Video>,
    val seedChannel: String? = null,
)

/**
 * בניית המיקסים.
 *
 * ## איך זה עובד אצל ספוטיפיי, ומה מזה אפשר כאן
 * "Daily Mix" של ספוטיפיי בנוי משני דברים: אשכולות של אמנים שנשמעים יחד
 * אצל אותו מאזין, ודגימה מתוך כל אשכול. את האשכולות הם לומדים ממיליוני
 * משתמשים — מידע שאין לנו, ובכוונה: כאן שום דבר לא עוזב את המכשיר.
 *
 * מה שכן יש הוא ההיסטוריה של המשתמש עצמו, הלייקים שלו והקטגוריות של
 * הרשימה הלבנה. משלושת אלה אפשר לבנות בדיוק את אותם סוגי מיקסים —
 * "האמן שלך", "הסגנון שלך", "עוד כמו מה ששמעת" — בלי שרת ובלי פרופיל
 * שנשמר אצל מישהו.
 *
 * ## למה מיקס חייב מינימום שירים
 * רשימת השמעה בת שלושה שירים היא לא רשימה, היא תזכורת. מיקס שלא מגיע
 * ל-[MIN_SONGS] פשוט לא נוצר — עדיף פחות מיקסים טובים מהרבה ריקים.
 */
object MixBuilder {

    private const val MIN_SONGS = 6
    private const val MAX_SONGS = 40
    private const val MAX_ARTIST_MIXES = 4

    /** כמה אמנים מובילים נחשבים "הטעם" לצורך מיקס יומי. */
    private const val CORE_ARTISTS = 6

    fun build(
        pool: List<Video>,
        likes: List<Video>,
        history: List<Video>,
        profile: TasteProfile,
        categoryOf: (String) -> String?,
        channelName: (String) -> String?,
        now: Long = System.currentTimeMillis(),
    ): List<Mix> {
        val all = (pool + likes + history).distinctBy { it.id }.filter { it.id.isNotBlank() }
        if (all.isEmpty()) return emptyList()

        val mixes = ArrayList<Mix>()
        val heard = (likes + history).mapTo(HashSet()) { it.id }

        // ── 1. המיקס היומי ────────────────────────────────────────────────
        // מהאמנים המובילים של המשתמש, מעורבב. זה המקבילה ל-Daily Mix:
        // לא הכי חדש ולא הכי אהוב, אלא חתך מייצג של הטעם.
        val core = profile.channelAffinity.entries
            .sortedByDescending { it.value }
            .take(CORE_ARTISTS)
            .map { it.key }
            .toSet()
        if (core.isNotEmpty()) {
            val daily = all.filter { it.channelId in core }
                .shuffledStable(now / 86_400_000)   // מתחלף פעם ביום
                .take(MAX_SONGS)
            if (daily.size >= MIN_SONGS) {
                mixes += Mix(
                    id = "daily",
                    title = "המיקס היומי שלך",
                    subtitle = "מהאמנים שאתה הכי שומע",
                    songs = daily,
                    seedChannel = core.firstOrNull(),
                )
            }
        }

        // ── 2. מיקס לכל אמן מוביל ─────────────────────────────────────────
        profile.channelAffinity.entries
            .sortedByDescending { it.value }
            .take(MAX_ARTIST_MIXES)
            .forEach { (channelId, _) ->
                val name = channelName(channelId) ?: return@forEach
                val songs = all.filter { it.channelId == channelId }
                    .shuffledStable(channelId.hashCode().toLong())
                    .take(MAX_SONGS)
                if (songs.size >= MIN_SONGS) {
                    mixes += Mix(
                        id = "artist_$channelId",
                        title = "מיקס $name",
                        subtitle = "${songs.size} שירים",
                        songs = songs,
                        seedChannel = channelId,
                    )
                }
            }

        // ── 3. מיקס לפי סגנון ─────────────────────────────────────────────
        profile.categoryAffinity.entries
            .sortedByDescending { it.value }
            .take(3)
            .forEach { (category, _) ->
                val label = categoryLabels[category] ?: return@forEach
                val songs = all.filter { categoryOf(it.channelId) == category }
                    .shuffledStable(category.hashCode().toLong())
                    .take(MAX_SONGS)
                if (songs.size >= MIN_SONGS) {
                    mixes += Mix(
                        id = "cat_$category",
                        title = label,
                        subtitle = "הסגנון שאתה הכי שומע",
                        songs = songs,
                        seedChannel = songs.firstOrNull()?.channelId,
                    )
                }
            }

        // ── 4. גילויים ────────────────────────────────────────────────────
        // מה שעוד לא שמעת, מקטגוריות שכן אהובות עליך. בלי הקטע הזה מיקס
        // אישי הופך למעגל סגור שמחזיר תמיד את אותם שירים.
        val discovery = all.asSequence()
            .filter { it.id !in heard }
            .filter { (profile.categoryAffinity[categoryOf(it.channelId)] ?: 0.0) > 0.15 }
            .sortedByDescending { it.publishedAt }
            .take(MAX_SONGS)
            .toList()
        if (discovery.size >= MIN_SONGS) {
            mixes += Mix(
                id = "discovery",
                title = "גילויים",
                subtitle = "לא שמעת עדיין",
                songs = discovery,
                seedChannel = discovery.firstOrNull()?.channelId,
            )
        }

        // ── 5. שוב ושוב ───────────────────────────────────────────────────
        val repeats = likes.filter { it.id in history.mapTo(HashSet()) { h -> h.id } }
            .take(MAX_SONGS)
        if (repeats.size >= MIN_SONGS) {
            mixes += Mix(
                id = "repeat",
                title = "שוב ושוב",
                subtitle = "אהבת גם שמעת",
                songs = repeats,
                seedChannel = repeats.firstOrNull()?.channelId,
            )
        }

        return mixes
    }

    /**
     * ערבוב יציב לפי זרע.
     *
     * shuffle() רגיל מחזיר סדר אחר בכל composition, והמסך היה קופץ בכל
     * גלילה. זרע קבוע נותן סדר "מעורבב" שנשאר יציב — ומתחלף רק כשהזרע
     * מתחלף, למשל פעם ביום במיקס היומי.
     */
    private fun <T> List<T>.shuffledStable(seed: Long): List<T> =
        shuffled(java.util.Random(seed))
}

File: flutter_app/lib/data/mix_builder.dart
import '../models/channel.dart';
import '../models/video.dart';
import 'feed_ranker.dart';
import 'search_engine.dart';

/// מיקס מוכן להשמעה — כותרת, תת-כותרת ורשימת שירים.
class Mix {
  final String id;
  final String title;
  final String subtitle;
  final List<Video> songs;
  final String? seedChannel;

  const Mix({
    required this.id,
    required this.title,
    required this.subtitle,
    required this.songs,
    this.seedChannel,
  });

  String? get artwork => songs.isEmpty ? null : songs.first.thumbnailUrl;
}

/// בונה את המיקסים של FilterMusic ממה שהמשתמש באמת שומע.
///
/// הכל מקומי ובלי אף בקשת רשת: המיקס נגזר מהפיד שכבר נטען, מהלייקים
/// ומההיסטוריה. מיקס שמתבסס על חיפוש היה גם איטי וגם יקר במכסה, והוא
/// ממילא לא היה מדויק יותר.
class MixBuilder {
  MixBuilder._();

  static const int _minSongs = 6;
  static const int _maxSongs = 40;
  static const int _maxArtistMixes = 4;
  static const int _coreArtists = 6;

  static List<Mix> build({
    required List<Video> pool,
    required List<Video> likes,
    required List<Video> history,
    required TasteProfile profile,
    required String? Function(String) categoryOf,
    required String? Function(String) channelName,
  }) {
    final all = [...pool, ...likes, ...history]
        .where((v) => v.id.isNotEmpty)
        .fold<Map<String, Video>>({}, (acc, v) {
          acc.putIfAbsent(v.id, () => v);
          return acc;
        })
        .values
        .where((v) => kMusicCategories.contains(categoryOf(v.channelId)))
        .toList();
    if (all.isEmpty) return const [];

    final mixes = <Mix>[];
    final heard = {...likes.map((v) => v.id), ...history.map((v) => v.id)};

    // ── המיקס היומי ────────────────────────────────────────────────────
    // הערוצים שהמשתמש הכי קשור אליהם, בערבוב יומי קבוע — אותו מיקס לאורך
    // היום, חדש למחרת.
    final core = (profile.channelAffinity.entries.toList()
          ..sort((a, b) => b.value.compareTo(a.value)))
        .take(_coreArtists)
        .map((e) => e.key)
        .toSet();
    if (core.isNotEmpty) {
      final daily = all.where((v) => core.contains(v.channelId)).toList();
      if (daily.length >= _minSongs) {
        mixes.add(Mix(
          id: 'daily',
          title: 'המיקס היומי שלך',
          subtitle: 'נבנה ממה שאתה שומע — מתעדכן מעצמו',
          songs: shuffledStable(daily).take(_maxSongs).toList(),
        ));
      }
    }

    // ── מיקס לכל אמן ───────────────────────────────────────────────────
    for (final channelId in core.take(_maxArtistMixes)) {
      final name = channelName(channelId);
      if (name == null || name.isEmpty) continue;
      final songs = all.where((v) => v.channelId == channelId).toList();
      if (songs.length < _minSongs) continue;
      mixes.add(Mix(
        id: 'artist_$channelId',
        title: 'מיקס $name',
        subtitle: 'הכי מושמע אצלך',
        songs: shuffledStable(songs).take(_maxSongs).toList(),
        seedChannel: channelId,
      ));
    }

    // ── מיקס לפי סגנון ─────────────────────────────────────────────────
    final topCategories = (profile.categoryAffinity.entries.toList()
          ..sort((a, b) => b.value.compareTo(a.value)))
        .take(2);
    for (final entry in topCategories) {
      final label = categoryLabels[entry.key];
      if (label == null) continue;
      final songs =
          all.where((v) => categoryOf(v.channelId) == entry.key).toList();
      if (songs.length < _minSongs) continue;
      mixes.add(Mix(
        id: 'cat_${entry.key}',
        title: 'מיקס $label',
        subtitle: 'הסגנון שאתה חוזר אליו',
        songs: shuffledStable(songs).take(_maxSongs).toList(),
      ));
    }

    // ── גילוי ──────────────────────────────────────────────────────────
    // דווקא מה שעוד לא שמעת — אחרת המיקסים הופכים למעגל סגור.
    final discovery = all.where((v) => !heard.contains(v.id)).toList();
    if (discovery.length >= _minSongs) {
      mixes.add(Mix(
        id: 'discovery',
        title: 'גילויים',
        subtitle: 'מהערוצים המאושרים — מה שעוד לא שמעת',
        songs: shuffledStable(discovery).take(_maxSongs).toList(),
      ));
    }

    return mixes;
  }
}
